.travis.yml

```diff
@@ -33,6 +33,10 @@ matrix:
       env: TARGET=mips64-unknown-linux-gnuabi64 NORUN=1
     - name: "mips64el-unknown-linux-gnuabi64 - build-only"
       env: TARGET=mips64el-unknown-linux-gnuabi64 NORUN=1
+    - name: "mipsisa64r6el-unknown-linux-gnuabi64 - no assert_instr"
+      env: TARGET=mipsisa64r6el-unknown-linux-gnuabi64 STDSIMD_DISABLE_ASSERT_INSTR=1
```

Comment:
Why is `assert_instr` disabled for these ?

Reply:
Assert instruction is not usable because llvm optimizes builtins into other instructions. So, there is no one to one mapping between builtins and resulting instruction.
Yes, this is redundant (already added this into run.sh file) but I wanted to make obvious.
I can remove it.

Reply:
For which builtins does this happen and for which targets ?

Reply:
This happens for mipisa64r6 and mipisa64r6el targets, for following builtins:

```
    core_arch::mips::msa::assert___msa_aver_u_b_aver_u_b
    core_arch::mips::msa::assert___msa_aver_u_d_aver_u_d
    core_arch::mips::msa::assert___msa_aver_u_h_aver_u_h
    core_arch::mips::msa::assert___msa_aver_u_w_aver_u_w
    core_arch::mips::msa::assert___msa_bclr_d_bclr_d
    core_arch::mips::msa::assert___msa_bclri_b_bclri_b
    core_arch::mips::msa::assert___msa_binsli_b_binsli_b
    core_arch::mips::msa::assert___msa_binsli_d_binsli_d
    core_arch::mips::msa::assert___msa_binsli_h_binsli_h
    core_arch::mips::msa::assert___msa_binsli_w_binsli_w
    core_arch::mips::msa::assert___msa_binsri_b_binsri_b
    core_arch::mips::msa::assert___msa_binsri_d_binsri_d
    core_arch::mips::msa::assert___msa_binsri_h_binsri_h
    core_arch::mips::msa::assert___msa_binsri_w_binsri_w
    core_arch::mips::msa::assert___msa_bmnzi_b_bmnzi_b
    core_arch::mips::msa::assert___msa_bmz_v_bmz_v
    core_arch::mips::msa::assert___msa_bmzi_b_bmzi_b
    core_arch::mips::msa::assert___msa_bneg_d_bneg_d
    core_arch::mips::msa::assert___msa_bsel_v_bsel_v
    core_arch::mips::msa::assert___msa_bset_d_bset_d
    core_arch::mips::msa::assert___msa_clei_s_d_clei_s_d
    core_arch::mips::msa::assert___msa_clti_s_d_clti_s_d
    core_arch::mips::msa::assert___msa_copy_u_d_copy_u_d
    core_arch::mips::msa::assert___msa_dpadd_u_h_dpadd_u_h
    core_arch::mips::msa::assert___msa_ld_b_ld_b
    core_arch::mips::msa::assert___msa_ld_h_ld_h
    core_arch::mips::msa::assert___msa_ld_w_ld_w
    core_arch::mips::msa::assert___msa_ldi_d_ldi_d
    core_arch::mips::msa::assert___msa_ldi_h_ldi_h
    core_arch::mips::msa::assert___msa_ldi_w_ldi_w
    core_arch::mips::msa::assert___msa_maxi_s_d_maxi_s_d
    core_arch::mips::msa::assert___msa_mini_s_d_mini_s_d
    core_arch::mips::msa::assert___msa_ori_b_ori_b
    core_arch::mips::msa::assert___msa_slli_b_slli_b
    core_arch::mips::msa::assert___msa_srli_b_srli_b
    core_arch::mips::msa::assert___msa_st_b_st_b
    core_arch::mips::msa::assert___msa_st_h_st_h
    core_arch::mips::msa::assert___msa_st_w_st_w
```
Example:
```

disassembly for msa_st_w_shim_st_w: 
	 0: lui at,0x14 
	 1: ldi.b $w0,0 
	 2: daddiu v1,zero,42 
	 3: daddu at,at,t9 
	 4: insert.d $w0[0],a0 
	 5: daddiu at,at,-18472 
	 6: ld v0,-11128(at) 
	 7: ld at,-11136(at) 
	 8: insert.d $w0[1],a1 
	 9: daddiu v0,v0,-3997 
	10: sd v1,8(at) 
	11: sd v0,0(at) 
	12: daddiu at,a2,2044 
	13: jr ra 
	14: st.d $w0,0(at) 
thread 'core_arch::mips::msa::assert___msa_st_w_st_w' panicked at 'failed to find instruction `st.w` in the disassembly'
```

Reply:
So why isn't `st.w` being generated? Is the generated code more efficient for a CPU with MSA than the MSA instruction itself? Do we have a bug in the toolchain ? (In Rust, LLVM, etc.)

Reply:
I don't know if the generated code is more efficient (it should be the same), but it makes no difference on little-endian target. It's LLVM thing. Because of calling convention every vector gets passed the same as it would be v2i64, so I am guessing that this makes LLVM prefer st.d over st.w.

Reply:
> Because of calling convention every vector gets passed the same as it would be v2i64,

Is a different calling convention possible? The vector types appear to be properly defined (using different element types and number of elements). 

Looking at the LLVM intrinsics, they appear to be incorrectly declared to use the `"C"` ABI, and should use the unnamed ABI instead (e.g. here: https://github.com/rust-lang-nursery/stdsimd/blob/master/crates/core_arch/src/mips/msa.rs#L71). But I don't know if that will make a difference for these particular targets.

If that isn't causing this (or if that makes no sense), maybe `stdsimd-verify` should just check that the `prefix` in the `prefix.suffix` instruction names is found, instead of trying to match `prefix.suffix`.

Reply:
> If that isn't causing this (or if that makes no sense), maybe stdsimd-verify should just check that the prefix in the prefix.suffix instruction names is found, instead of trying to match prefix.suffix (maybe we should just change the assert_instr to only contain the prefix).

This would help with st.w but the other issues.
Some of the other issues https://github.com/Microsoft/llvm/blob/master/lib/Target/Mips/MSA.txt
I think that this would not be possible to reliably assert every builtin on all optimization levels.
x86 has done the same for AVX.

Reply:
> I think that this would not be possible to reliably assert every builtin on all optimization levels.

`assert_instr` is only checked in `--release`

Reply:
If  `prefix` would be used instead of `prefix.suffix `in `assert_instr`, release mode build would be successful, but the problem would appear in stdsimd-verify:
https://github.com/rust-lang-nursery/stdsimd/blob/master/crates/stdsimd-verify/tests/mips.rs#L315-L325

Reply:
So maybe the best we can do is, for mips targets, in the `assert_instr` macro, only check the prefix. That way, the prefix.suffix is kept in the `assert_instr` annotation, and `stdsimd-verify` has all the information. 

That could happen somewhere around here: https://github.com/rust-lang-nursery/stdsimd/blob/master/crates/stdsimd-test/src/lib.rs#L88

Reply:
There is another problem. For some builtins, even this is not enough.
For example, for a binsli.d builtin, the situation is as follows:

```
disassembly for msa_binsli_d_shim_binsli_d: 
	 0: ldi.b $w0,0 
	 1: lui at,0x14 
	 2: daddiu a1,zero,58 
	 3: insert.d $w0[0],a2 
	 4: daddu at,at,t9 
	 5: daddiu at,at,-4200 
	 6: insert.d $w0[1],a3 
	 7: ld a0,-10752(at) 
	 8: ld at,-10744(at) 
	 9: copy_s.d v0,$w0[0] 
	10: copy_s.d v1,$w0[1] 
	11: daddiu at,at,22660 
	12: sd a1,8(a0) 
	13: jr ra 
	14: sd at,0(a0) 
thread 'core_arch::mips::msa::assert___msa_binsli_d_binsli_d' panicked at 'failed to find instruction `binsli.d` in the disassembly'
```
The `prefix` does not appear, so, checking only `prefix` will not solve the problem for this situation.
Most of the problematic builtins are listed here:
https://github.com/Microsoft/llvm/blob/master/lib/Target/Mips/MSA.txt

Reply:
So what we do for other targets here, is comment the `assert_instr` out, and white-list those in `stdsimd-verfiy` as "its ok that this does not have an assert_instr` because of X". 

The comment that comments the `assert_instr` out, should explain why: is this an LLVM bug (link to the bug) ? Is LLVM generating more optimized code here? Then explain why this code is more optimized, add an assert_instr for the more optimized version, and white-list that in stdsimd-verify, etc. 

---

Off-topic: it's a bit off-topic, but I think that when this PR is merged, we can start talking about stabilizing this intrinsics. That might need a small RFC, or an FCP, but I can help with that and with pushing this through the process. You have done an awesome job here, implementing all intrinsics, adding tests, automatic-verification, etc. I've barely have had to help you, and I think when this is stabilized, maybe it would be worthy of being part of the rust blog release announcement to explain what you did, how you did it, how much time it costed, what were the issues that we could do better process wise, etc. Right now, we only have x86 intrinsics, and showing that we are committed to shipping intrinsics for other archs like WASM, MIPS, ARM, etc. would be a good thing IMO.